### ISSUE TITLE
SSL_CTX_free - Access violation

### ISSUE THREAD

**Issue by amaljesudas** [openssl/openssl#18579]
Hi,

I just encountered a bizarre scenario in a sample client application I was trying out.
Was testing the client against a server using TLS.
A third party library handles the client side communication and this library uses the secure functions we provide to establish secure communication.
I have implemented the secure functions using openssl [Openssl version is 1.1.1l].
The TLS option of my client can be switched ON and OFF, meaning it can have both secure(TLS) and non-secure communication with servers.
For once, after making a successful client-server TLS communication, I stopped my server, switched OFF the TLS option at my client side and tried to establish a connection with the now non-existent server.
To my surprise, I got the below access violation:
**First-chance exception at 0x7B8AB83E (libcrypto-1_1.dll) in Client.exe: 0xC0000005: Access violation reading location 0xABABABAB.**
After some basic analysis, it was found that this happened when the function SSL_CTX_free was getting called.

I know that this is very limited information to make a query, but I am just looking for some suggestions or early leads.
Can someone tell me what could have possibly gone wrong?
Is there a scenario where the context of previous TLS connection gets reused? If yes, I have not done anything intentionally for this.
Any possible method to check SSL_CTX object before calling  SSL_CTX_free()? I have done the null check but want to know if there is any other validation that could be done before calling SSL_CTX_free().

Thanks in advance and sorry for the limited info I could furnish.

Regards,
Amal

**Comment by mattcaswell** [comment]
The location `0xABABABAB` is interesting. See:

https://www.asawicki.info/news_1292_magic_numbers_in_visual_c

> ABABABAB - Used by Microsoft's HeapAlloc() to mark "no man's land" guard bytes after allocated heap memory

So this is heap memory which is not allocated to the application but is nonetheless being read. I guess you need to figure out why. Perhaps a double free, or an attempt to free an uninitialised value?

**Comment by amaljesudas** [comment]
Hi @mattcaswell 

Thanks for the reply.
In fact the location is not a constant, although once in a while it does show up.
Below is the error I got on 5 iterations of the same steps to reproduce the issue:
**First-chance exception at 0x7B0C8C4E (libcrypto-1_1.dll) in Client.exe: 0xC0000005: Access violation reading location 0x0000002C.
First-chance exception at 0x773AFEE5 (ntdll.dll) in Client.exe: 0xC0000005: Access violation writing location 0x00000004.
First-chance exception at 0x7AFFB83E (libcrypto-1_1.dll) in Client.exe: 0xC0000005: Access violation reading location 0xABABABAB.	
First-chance exception at 0x773AFEE5 (ntdll.dll) in Client.exe: 0xC0000005: Access violation writing location 0x00000004.
First-chance exception at 0x7B0C8C4E (libcrypto-1_1.dll) in Client.exe: 0xC0000005: Access violation reading location 0x3E7B8FD9.**

The API call remains the same SSL_CTX_free.

In between, I was wondering if after calling SSL_CTX_free, do we need to set the SSL_CTX parameter to NULL?
Could see this done in some of the samples available online.

**Comment by mattcaswell** [comment]
> The API call remains the same SSL_CTX_free.

Nonetheless the problem is likely to be the same, i.e. attempting to use an object that has already been freed, or possibly using an object pointer that has not been initialised correctly.

> In between, I was wondering if after calling SSL_CTX_free, do we need to set the SSL_CTX parameter to NULL?

There isn't a simple yes/no answer to this question - which is more about C than OpenSSL specifically.

`SSL_CTX` objects are reference counted. When you call `SSL_CTX_new()` you get an object back with 1 reference. If you call `SSL_CTX_up_ref()` then it will increment the reference count. When you call `SSL_CTX_free` it will decrement the reference count. If the reference count drops to zero then the object will be freed and the memory that it was using will be released back to the OS. Any subsequent attempt to use the `SSL_CTX` after that point is likely to cause a crash.

In other words there must be exactly one call to `SSL_CTX_free()` for each call to `SSL_CTX_new()` or `SSL_CTX_up_ref()`.

You must not use an `SSL_CTX` once it has been freed (the ref count is zero). In order to avoid accidental use many applications will set the pointer to NULL after it has been freed. This is not strictly necessary if your application will never make use of the pointer again anyway - but is often considered good practice.

**Comment by amaljesudas** [comment]
@mattcaswell 

Yes, I get it.
As I mentioned in the first post, the actual handling of the communication is done by a third party library.
It uses the secure functions that we provide to establish secure TLS communication.
So we need to provide a separate function to create the contexts, separate functions to perform read/write operation and finally one separate function to do the cleanup.
It is in the clean up function that we free up the context.
Expectation is that the create context function is invoked first followed by read/write functions and finally the clean up function.
In issue scenario, as per logs, what I am seeing is that the clean up function is getting called without invoking the create context function first.
I am not sure if this is an issue with the third party library logic, but am assuming that an unnecessary SSL_CTX_free is being done even when a SSL_CTX_new has not happened.

**Comment by mattcaswell** [comment]
> In issue scenario, as per logs, what I am seeing is that the clean up function is getting called without invoking the create context function first.
I am not sure if this is an issue with the third party library logic, but am assuming that an unnecessary SSL_CTX_free is being done even when a SSL_CTX_new has not happened.

Right - yeah if you call `SSL_CTX_free` with a bad pointer (i.e. you didn't call SSL_CTX_new first) then this will definitely cause this crash. Make sure you don't do that :-)

**Comment by amaljesudas** [comment]
Thanks @mattcaswell
